### ISSUE TITLE
ACs for MVP

### ISSUE THREAD

**Issue by Tamsen1995** [Tamsen1995/patient-portal#6]
## Feature Development Acceptance Criteria: Patient List and Detail Page
- [x] AC1: The platform must display a list of patients.
- [x] AC2: Each patient in the list must link to a dedicated page containing the patient's summary.
- [x] AC3: The patient summary must include the patient's name, first name, age, height, weight, and gender.

## Highlights and Action Suggestions
- [ ] AC4: Each patient's page must have a section for highlights and action suggestions.
- [ ] AC5: The system must prompt to take a temperature reading for the current day if not already done.
- [ ] AC6: The system must prompt to administer medication for the current day if not already done.

## Historical Data and Interactions
- [x] AC7: Display a temperature chart with options to change the scale to 1/3/6 months.
- [x] AC8: List current medication, including name, dosage, start date, and end date.
- [x] AC9: Allow users to add or modify temperature records (maximum once per day).
- [x] AC10: Allow users to add or modify medication details.
- [ ] AC11: Allow users to indicate that a dose of medication has been taken on a given date.
- [ ] Note: Users can only add temperature for the current day, and only once per day. If there is already a temp for the current day, prevent adding another.

## Planned Enhancements Acceptance Criteria: Data Expansion
- [ ] AC12: Enable expansion of entered data to include medical history, allergies, etc.
- [ ] AC13: The platform should provide a comprehensive view of the patient's medical history.

## Real-time Data Integration
- [ ] AC14: Integrate the ability to automatically receive real-time information (e.g., blood pressure) from connected devices.

## AI Model Integration
- [ ] AC15: Implement an AI model trained on medical data to provide relevant suggestions based on the patient's condition and real-time data.

## Alert System
- [ ] AC16: Implement an alert system that triggers notifications if a vital sign falls below specified thresholds.

## Production and Long-term Solution Criteria: Production Readiness
- [x] AC17: Identify and outline missing elements for production readiness.
- [x] AC18: Propose an action plan addressing the missing elements for going live.

## Long-term Enhancement Plan
- [x] AC19: Propose a clear technical roadmap for implementing the planned enhancements.
- [x] AC20: Describe the infrastructure and specify GCP services for deploying the system, including enhancements.

## General and Technical Requirements
- [ ] AC21: Ensure the application's design is comprehensive, qualitative, and relevant.
- [x] AC22: Use patient data from the provided JSON file for the database.
- [ ] AC23: Propose a simple method to switch between patients for demonstration purposes.
- [ ] AC24: Ensure the application can run with only Docker installed, as per README instructions.

## Development and Documentation Criteria: Code and Data Structure Quality
- [ ] AC25: Maintain high code quality and correctness in data structure.
- [ ] AC26: Optimize operational efficiency of the platform.
- [x] AC27: Provide clear and understandable documentation, including responses to questions and a clear README file.

**Comment by Tamsen1995** [comment]
### AC19
### AC20


To deploy our medical platform on Google Cloud Platform (GCP), I have designed an infrastructure that supports scalability, reliability, and security, catering to the healthcare industry's stringent requirements. This deployment strategy encompasses both the current application features and the planned enhancements, ensuring a comprehensive and future-proof solution.

### Core Infrastructure

1. **Compute Engine:** For hosting our backend and frontend applications, I'll use GCP's Compute Engine to create and manage virtual machines (VMs) that scale according to the application's demand. This flexibility allows us to efficiently manage resource allocation, ensuring that the application remains responsive under varying loads.

2. **Cloud SQL:** Patient data, including health metrics and medication schedules, will be stored in Cloud SQL, GCP's fully-managed relational database. It offers high performance, scalability, and convenience for application data storage, with support for automated backups, replication, and encryption to ensure data security and compliance.

3. **Cloud Storage:** For storing static assets such as images and patient data files, I'll use Cloud Storage. It provides a secure and highly available object storage solution, which is essential for handling medical images and documents securely.

### Data Processing and Analytics

1. **BigQuery:** To analyze large datasets of patient health records for insights and reporting, I'll leverage BigQuery, GCP's serverless, highly scalable, and cost-effective multi-cloud data warehouse. It will support our AI-driven analysis and decision-making processes.

2. **Dataflow:** For real-time data processing, especially from wearable devices transmitting health metrics, Dataflow will be used. It allows us to process streams of data in real-time, providing up-to-date information for patient monitoring and alerting.

### AI and Machine Learning

1. **AI Platform:** The AI Platform will serve as the backbone for developing, training, and deploying our machine learning models, including the LSTM models for health suggestion predictions. This integrated tool facilitates the creation of sophisticated AI-driven features while managing them efficiently across the platform.

2. **Cloud AutoML:** To further enhance our platform with AI capabilities without deep machine learning expertise, Cloud AutoML will allow us to train high-quality custom models with minimal effort. This will be particularly useful for interpreting medical images or predicting health outcomes based on patient data.

### Security and Management

1. **Cloud Identity & Access Management (IAM):** Ensuring that only authorized users can access specific resources, Cloud IAM will provide fine-grained access control and visibility for managing permissions securely.

2. **Cloud Armor:** To protect our web applications from DDoS attacks, SQL injection, and other web vulnerabilities, Cloud Armor will be deployed. It ensures our application's and patients' data security.

3. **Cloud Healthcare API:** Integrating the Cloud Healthcare API will facilitate interoperability with healthcare applications, enabling secure data exchange and scalability in managing healthcare data, including HL7, FHIR, and DICOM standards for medical data.

### Monitoring and Operations

1. **Cloud Monitoring and Cloud Logging:** These services will provide comprehensive visibility into the health, performance, and logs of applications and infrastructure. Monitoring will help in setting up alerts based on specific metrics, while logging will capture and analyze logs from all components of the application.

2. **Operations Suite:** For managing and monitoring the health of applications, the Operations Suite (formerly Stackdriver) will be used. It combines monitoring, logging, error reporting, and tracing, providing a unified toolset to manage system performance and reliability.

**Comment by Tamsen1995** [comment]
### AC17: Identifying Missing Elements for Production Readiness

1. **Security Compliance and Data Protection:** 
   - Ensuring compliance with healthcare regulations such as HIPAA (Health Insurance Portability and Accountability Act) in the US or GDPR (General Data Protection Regulation) in Europe.
   - Implementing end-to-end encryption for data at rest and in transit.

2. **Disaster Recovery and Data Backup:**
   - Lacking a comprehensive disaster recovery plan.
   - Insufficient data backup strategies for patient data and application state.

3. **Scalability and Performance:**
   - Need for a scalable infrastructure that automatically adjusts to varying loads.
   - Performance optimization for latency-sensitive operations.

4. **Monitoring, Logging, and Alerting:**
   - Inadequate monitoring and logging mechanisms for early detection of issues.
   - Lack of an integrated alerting system for operational and security incidents.

5. **User Authentication and Access Control:**
   - Robust user authentication mechanisms including multi-factor authentication (MFA).
   - Fine-grained access control policies to ensure users have access only to the data and actions necessary for their role.

6. **Patient Data Validation and Integrity:**
   - Mechanisms to ensure the accuracy, completeness, and consistency of patient data.

7. **Third-party Integration and API Security:**
   - Security assessment and integration testing with third-party services, especially those handling sensitive medical data.
   - Secure API endpoints against common vulnerabilities.

8. **Legal and Regulatory Considerations:**
   - Finalizing agreements and compliance checks with legal counsel regarding data handling, privacy, and service level agreements (SLAs).

**Comment by Tamsen1995** [comment]
### AC18: Action Plan for Addressing Missing Elements

**Security Compliance and Data Protection:**
- Conduct a comprehensive security audit against HIPAA or GDPR requirements.
- Implement AES encryption for data at rest and TLS 1.3 for data in transit.

**Disaster Recovery and Data Backup:**
- Develop and implement a disaster recovery plan that includes regular, encrypted backups stored in multiple geographical locations.
- Test the disaster recovery plan quarterly.

**Scalability and Performance:**
- Utilize GCP's managed services like App Engine or Kubernetes Engine for automatic scaling.
- Conduct performance testing to identify bottlenecks and optimize critical paths in the application.

**Monitoring, Logging, and Alerting:**
- Implement GCP's Operations Suite for comprehensive monitoring, logging, and real-time alerting.
- Set up dashboards for key metrics and alerts for anomalies.

**User Authentication and Access Control:**
- Integrate an identity provider that supports MFA and robust authentication protocols.
- Use GCP Cloud IAM for detailed access control policies and audit logging.

**Patient Data Validation and Integrity:**
- Implement input validation at both the frontend and backend to ensure data integrity.
- Schedule regular data quality audits and implement automated data cleaning processes.

**Third-party Integration and API Security:**
- Perform security assessments for all third-party services and ensure they comply with necessary standards.
- Secure APIs using OAuth 2.0, and regularly update and audit API access policies.

**Legal and Regulatory Considerations:**
- Collaborate with legal experts to ensure all data handling practices are in line with current laws and regulations.
- Prepare and review all necessary documentation and SLAs for compliance and legal obligations.
